### PR TITLE
numerical LU factorization

### PR DESCRIPTION
**Pull request by upsj**
This adds numerical LU factorization kernels to initialize the factors from the matrix and run the factorization. It uses the Csr lookup structures to implement the required sparse vector additions, and a sync-free waiting approach similar to our custom triangular solvers.

### CODE REVIEW

File: common/cuda_hip/factorization/lu_kernels.hpp.inc
Comment (fritzgoebel) [inline common/cuda_hip/factorization/lu_kernels.hpp.inc:59]:
```suggestion
    for (auto nz = factor_begin + lane; nz < factor_end; nz += config::warp_size) {
```
I think this should be strided access with the warp or did I miss something?

**Review by fritzgoebel — COMMENTED** [review]
Looks great, but omp and dpcpp are missing completely. Maybe add them as GKO_NOT_IMPLEMENTED?

File: common/cuda_hip/factorization/lu_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/factorization/lu_kernels.hpp.inc:59]: great catch!

File: common/cuda_hip/factorization/lu_kernels.hpp.inc
Comment (MarcelKoch) [inline common/cuda_hip/factorization/lu_kernels.hpp.inc:126]: would `__threadfence_block();` suffice here?

File: common/cuda_hip/factorization/lu_kernels.hpp.inc
Comment (MarcelKoch) [inline common/cuda_hip/factorization/lu_kernels.hpp.inc:193]: isn't `tmp_storage.resize_and_reset(2)` enough? You are only using the first and last entry, and ignoring everything in between.

**Review by MarcelKoch — APPROVED** [review]
Looks really good, especially the cuda/hip implementation. I still left a few questions there.
Besides that, I guess tests with edge cases (0x0 matrix, ...) will follow at a later point.

File: common/cuda_hip/factorization/lu_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/factorization/lu_kernels.hpp.inc:94]: CUDA doesn't guarantee that new blocks are scheduled in a monotonic order (even though some NVIDIA libraries like cub rely on it), so just to be safe, we added this workaround. I'm also pretty annoyed with them for this behavior, but it might change due to more complex scheduling on multi-chiplet GPUs or so, I could imagine? Anyways, without this monotonicity, we could livelock by starving previous rows of compute time.

File: common/cuda_hip/factorization/lu_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/factorization/lu_kernels.hpp.inc:126]: probably, yes

File: common/cuda_hip/factorization/lu_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/factorization/lu_kernels.hpp.inc:193]: we use tmp_storage for two things: the atomic counter and 0/1 flags for each row

Comment (MarcelKoch) [reply common/cuda_hip/factorization/lu_kernels.hpp.inc:193]: Oh, I see, I didn't realize that the `store` does the indexing internally.

File: common/cuda_hip/components/volatile.hpp.inc
Comment (thoasm) [inline common/cuda_hip/components/volatile.hpp.inc:55]: I would prefer if you could have the SFINAE in the return type (`void`) of the `store` function, and not in the argument type itself (especially for readability reasons).

File: reference/test/factorization/lu_kernels.cpp
Comment (thoasm) [inline reference/test/factorization/lu_kernels.cpp:76]: Requires header `<fstream>` (which is currently not included).

File: reference/test/factorization/lu_kernels.cpp
Comment (thoasm) [inline reference/test/factorization/lu_kernels.cpp:102]: Requires both `<ginkgo/core/base/types.hpp>` and `<ginkgo/core/base/array.hpp>`

File: test/factorization/lu_kernels.cpp
Comment (thoasm) [inline test/factorization/lu_kernels.cpp:90]: Missing header `<fstream>`

File: test/factorization/lu_kernels.cpp
Comment (thoasm) [inline test/factorization/lu_kernels.cpp:133]: Missing header `ginkgo/core/base/types.hpp` and `ginkgo/core/base/array.hpp`

**Review by thoasm — APPROVED** [review]
LGTM!

**Comment by thoasm** [comment]
Nit: maybe add a short description of this PR (before you merge).

**Comment by upsj** [comment]
I'll wait for the pipelines to finish here, then rebase and merge directly, since the previous changes only related to Github Actions.

**Comment by upsj** [comment]
The clang-cuda issues seem unrelated, so I'll still proceed

**Comment by upsj** [comment]
rebase!

**Pull request merged**
